src/statemachine/statemachine.js

```diff
@@ -0,0 +1,330 @@
+         * @remarks It's not normally necessary to call this function as states are usually defined
+         *     up front by passing a state descriptor to the constructor
+         * @example
+         * //Add a die method to the class and define it for the 'walking' state
+         * this.stateMachine.define('walking', 'die', function() { //do something });
+         */
+        define: function (state, method, fn) {
+            var registrationList = [];
+            var registerState = function registerState(stateName, definition) {
+                forEach(definition, function (method, methodName) {
+                    if (typeof method !== 'function') {
+                        return;
+                    }
+                    if (methodName != "enter" && methodName != "exit") {
+                        this.states[stateName][methodName] = method;
+                        registrationList.push(methodName);
+                    }
+                    else if (methodName == "enter") {
+                        this.enter[stateName] = this.enter[stateName] || [];
+                        this.enter[stateName].push(method);
+                    }
+                    else if (methodName == "exit") {
+                        this.exit[stateName] = this.exit[stateName] || [];
+                        this.exit[stateName].push(method);
+                    }
+                }.bind(this));
+            }.bind(this);
+
+            if (typeof state == 'object' && !method && !fn) {
+                forEach(state, function (s, i) {
+                    this.states[i] = this.states[i] || {};
+                    registerState(i, s);
+                }.bind(this));
+                this._register(registrationList);
+                return this;
+            }
+            else if (typeof method == 'object' && typeof state == 'string' && fn) {
+                this.states[state] = this.states[state] || {};
+                registerState(state, method);
+                this._register(registrationList);
+                return this;
+            }
+            else if (typeof method == 'string' && typeof state == 'string' && fn) {
+                this.states[state] = this.states[state] || {};
+                this.states[state][method] = fn;
+                this._register(method);
+                return this;
+            }
+            else {
+                throw new Error("Unrecognized configuration");
+            }
+        },
+        _register: function register(method, decorate) {
+
+            decorate = decorate || this._decorate;
+            method = Array.isArray(method) ? method : method.split(' ');
+            method.filter(function (m) {
+                return m;
+            }).forEach(function (method) {
+                var working = noop;
+                var lastState;
+                this.change(function () {
+                    this.states[lastState = this.state] = this.states[this.state] || {};
```

Comment:
`this.states[this.state]` is referenced many times, having a local variable `var state = this.states[this.state];` and then referencing it later will be a bit cleaner.